File: src/main/java/legacy/DiscountStrategy.java
package legacy;

import java.math.BigDecimal;

public interface DiscountStrategy {

    Money calculateDiscountedValue(Money amount);

    class Fixed implements DiscountStrategy {

        private final Money value;

        public Fixed(Money value) {
            this.value = value;
        }

        @Override
        public Money calculateDiscountedValue(Money amount) {
            Money discountedValue = amount.minus(value);
            if(discountedValue.getValue().compareTo(BigDecimal.ZERO) > 0) {
                return discountedValue;
            } else {
                return Money.of(BigDecimal.ZERO, amount.getCurrency());
            }
        }

        public Money getValue() {
            return value;
        }
    }

    class Percentage implements DiscountStrategy {

        private final Percent value;

        public Percentage(Percent percentage) {
            this.value = percentage;
        }

        @Override
        public Money calculateDiscountedValue(Money amount) {
            return amount.minus(amount.times(value));
        }

        public Percent getValue() {
            return value;
        }
    }
}

File: src/main/kotlin/refactored/DiscountStrategy.kt
package refactored

import java.math.BigDecimal

sealed interface DiscountStrategy {
    fun calculateDiscountedValue(amount: Money): Money
    class Fixed(val value: Money) : DiscountStrategy {
        override fun calculateDiscountedValue(amount: Money): Money {
            val discountedValue = amount.minus(value)
            return if (discountedValue.value > BigDecimal.ZERO) {
                discountedValue
            } else {
                Money.Companion.of(BigDecimal.ZERO, amount.currency)
            }
        }
    }

    class Percentage(val value: Percent) : DiscountStrategy {

        override fun calculateDiscountedValue(amount: Money): Money {
            return amount.minus(amount.times(value))
        }
    }
}